File: apps/android/core/src/main/java/com/aatech/betweenus/core/store/ChannelFocus.kt
package com.aatech.betweenus.core.store

import com.aatech.betweenus.core.data.PresenceSocket
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch

/**
 * Telling the server which conversation is on this screen.
 *
 * The other half of the rule [PushGate] applies locally, and the half only a
 * server can apply: a message in a channel somebody is reading on their laptop
 * should not wake the phone in their pocket. `notification-service` asks
 * presence-service who has a channel focused before it fans a push out - see
 * `push-suppression.md`.
 *
 * Two facts make a focus, and both already exist here: which conversation is
 * open ([Conversation.visibleChannelId]) and whether the app is in front of
 * anybody ([AppForeground]). A phone locked mid-conversation is not somebody
 * reading it - the chat screen is still composed behind the lock screen, which
 * is exactly the case that would otherwise go silent forever.
 *
 * ponytail: no lifecycle observers of its own. Both facts are already tracked
 * for the local rule; this only reports them.
 */
object ChannelFocus {

    /**
     * Well inside presence-service's 90-second staleness window. The socket's
     * own heartbeat refreshes the entry too; this covers a reconnect that
     * landed on a different presence instance.
     */
    private const val HEARTBEAT_MS = 60_000L

    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Default)

    /** What the server has been told, so nothing is said twice. */
    @Volatile
    private var reported: String? = null

    private var started = false

    fun init() {
        if (started) return
        started = true

        // Re-asserted on every reconnect, like the status and the call join: the
        // server keeps nothing across connections, and a focus that is not
        // re-sent ages out into a phone that starts buzzing for a conversation
        // that is still on screen.
        PresenceSocket.onReconnect = {
            reported = null
            apply()
        }

        scope.launch {
            while (true) {
                delay(HEARTBEAT_MS)
                // Re-sent rather than skipped when nothing has changed: this is
                // the heartbeat, and saying the same thing again is its job.
                reported?.let { PresenceSocket.focus(it) }
            }
        }
    }

    /**
     * Called whenever either half of the fact changes - a channel opened or
     * closed, or the app coming to the front or going away.
     */
    fun apply() {
        val next = if (AppForeground.visible) Conversation.visibleChannelId else null
        val previous = reported
        if (next == previous) return

        // The old channel is blurred rather than left to expire: ninety seconds
        // of a stale focus is ninety seconds of missed notifications for a
        // conversation this phone has already left.
        if (previous != null) PresenceSocket.blur(previous)
        if (next != null) PresenceSocket.focus(next)
        reported = next
    }

    /** Signing out: this account is not reading anything any more. */
    fun forget() {
        reported?.let { PresenceSocket.blur(it) }
        reported = null
    }
}

File: apps/desktop/src/services/channel-focus.ts
/**
 * Telling the server which conversation is in front of this person.
 *
 * It exists for one thing: so a phone is not woken for a message its owner is
 * reading on this screen. `notification-service` asks presence-service who has
 * a channel focused before it fans a push out, and drops them - see
 * `push-suppression.md`.
 *
 * "Focused" is deliberately strict. A channel is reported only while it is the
 * open one *and* this window has the operating system's focus. A desktop left
 * open on #general behind a browser is not somebody reading #general, and
 * treating it as one would silently swallow their notifications for the rest of
 * the day - which is a far worse failure than one redundant buzz.
 *
 * The same module serves the browser build, because it is the same app. A tab
 * gets `visibilitychange` as well, which is the closer thing a tab has to being
 * looked at.
 */
import { useChatStore } from '../stores/chat';
import { presenceSocket } from './socket';

/**
 * Well inside presence-service's 90-second staleness window, so a focus that
 * is still true is never allowed to age out of Redis. The socket's own
 * heartbeat refreshes it too; this covers a client that reconnected onto a
 * different presence instance.
 */
const HEARTBEAT_MS = 60_000;

export function startChannelFocus(): () => void {
  /** What the server has been told, so nothing is said twice. */
  let reported: string | null = null;
  let timer: ReturnType<typeof setInterval> | null = null;

  const focused = (): boolean => document.hasFocus() && !document.hidden;

  const current = (): string | null => {
    if (!focused()) return null;
    return useChatStore.getState().activeChannelId;
  };

  const apply = (): void => {
    const next = current();
    if (next === reported) return;

    // Blur the old one explicitly rather than letting it expire. Ninety
    // seconds of a stale focus is ninety seconds of missed notifications for a
    // channel this window has already left.
    if (reported) presenceSocket.send({ type: 'channel.blur', channelId: reported });
    if (next) presenceSocket.send({ type: 'channel.focus', channelId: next });
    reported = next;
  };

  const unsubscribe = useChatStore.subscribe(apply);
  window.addEventListener('focus', apply);
  window.addEventListener('blur', apply);
  document.addEventListener('visibilitychange', apply);

  timer = setInterval(() => {
    // Re-sent rather than skipped when nothing has changed: this is the
    // heartbeat, and its whole job is to say the same thing again.
    if (reported) presenceSocket.send({ type: 'channel.focus', channelId: reported });
  }, HEARTBEAT_MS);

  apply();

  return () => {
    if (timer) clearInterval(timer);
    unsubscribe();
    window.removeEventListener('focus', apply);
    window.removeEventListener('blur', apply);
    document.removeEventListener('visibilitychange', apply);
    if (reported) presenceSocket.send({ type: 'channel.blur', channelId: reported });
    reported = null;
  };
}
